lib/fs/walkfs.go

```diff
@@ -55,6 +104,13 @@ func (f *walkFilesystem) walk(path string, info FileInfo, walkFn WalkFunc) error
 		return nil
 	}
 
+	if ancestors.AppendUnlessPresent(info) {
+		defer ancestors.Remove(info)
+	} else {
+		l.Warnf("Infinite filesystem recursion detected on path '%s', not walking further down", path)
```

Comment:
This check won't protect against scanning files several iterations deep into a junction loop, if such a path comes in via events or API. (E.g. if you have `c:\foo\bar -> c:\foo` and we ask for a scan of `c:\foo\bar\bar\bar\bar\somefile`.)

If we don't get fs notifications like that (because Windows is smart and canonicalizes the name before notifying or the fsnotify package does) I don't think it's a big deal. But if we do get notifications like that if you open `c:\foo\bar\bar\bar\bar\somefile` in notepad I think it's a problem.

Reply:
Can you explain a bit more what you mean? I am aware that if, e.g. in a Linux copy,  you place a file into the empty folder that was synchronized from a directory junction in infinite recursion, it would synchronize back. But IMO this shouldn't hurt. This protection is meant for protecting Syncthing from crashing, nothing more. But maybe I overlooked something?

_Example:_
Wiindows: directory junction `c:\foo\bar -> c:\foo` would synchronize `c:\foo\bar\bar` to Linux, and then stop. At Linux side, `c:\foo\bar\bar` would be an empty folder. If you create `c:\foo\bar\bar\dummy,txt` at the Linux side, it would synchronize back to Windows to `c:\foo\bar\dummy.txt`. I don't think this is a problem. Is it?

Reply:
It is, because on the Windows side we now have two file entries referring to the same physical file, any change to it will be sent as a change to both foo/bar/dummy and foo/bar/bar/dummy because we know about it under both names. Incoming changes from linux would also change the "other" file, and vice versa. This is essentially a variant of the aliasing that we get with case-only renamed between linux/windows, and it's havoc that we rather avoid.

Reply:
I still don't understand, because what you describe is expected. If you had `c:\foo\bar -> c:\baz` and `c:\foo\bar2 -> c:\baz` at Windows side, you would still have at Linux side two corresponding directories: `.../foo/bar` and `.../foo/bar2`. I had tested that if you change something in `.../foo/bar`, it gets propagated to `c:\baz`, but not to `.../foo/bar2`.

Reply:
> expected

Mmm. I guess this might fall under "don't do that, then", yeah.